### PR TITLE
[v6r15] Fixes for GFAL2

### PR DESCRIPTION
**Pull request by chaen**
This contains few fixes in the gfal2 plugins, as well as a few adaptation of the StorageElement and StorageBase. Namely, getStorageParameters can now take a protocol rather than just a plugin name, which is very helpful when dealing with two plugins able to manage srm (GFAL2_SRM2 and SRM2). Hopefully, all this will become void when https://github.com/DIRACGrid/DIRAC/pull/2668 will finally be merged.

### CODE REVIEW

File: Resources/Storage/GFAL2_SRM2Storage.py [hunk 1]
```diff
@@ -162,7 +166,7 @@ def __getSingleTransportURL( self, path, protocols = False ):
         return S_ERROR( errStr )
     else:
       errStr = 'GFAL2_SRM2Storage.__getSingleTransportURL: %s' % res['Message']
-      return S_ERROR( errStr )
+      return res
```
Comment (fstagni) [inline Resources/Storage/GFAL2_SRM2Storage.py]: so now `errStr` is unused?

File: Resources/Storage/GFAL2_StorageBase.py [hunk 2]
```diff
@@ -267,37 +257,37 @@ def __putSingleFile( self, src_file, dest_url, sourceSize ):
...
+
+    # TODO: as soon as self.protocolParameters contains all known protocols to the SE
+    # we wont need this hard coded list below anymore but can implement a function in StorageBase
+    # similar to isNativeURL which returns true if the src_file contains a known protocol.
+    protocols = ['srm', 'root']
+    if any( src_file.startswith( protocol + ':' ) for protocol in protocols ):
+      src_url = src_file
+      if not sourceSize:
+        errStr = "GFAL2_StorageBase.__putFile: For file replication the source file size in bytes must be provided."
+        self.log.error( errStr, src_file )
```
Comment (fstagni) [inline Resources/Storage/GFAL2_StorageBase.py]: error code here?

File: Resources/Storage/GFAL2_StorageBase.py [hunk 3]
```diff
@@ -267,37 +257,37 @@ def __putSingleFile( self, src_file, dest_url, sourceSize ):
...
+        return S_ERROR( errStr )
```
Comment (fstagni) [inline Resources/Storage/GFAL2_StorageBase.py]: error code here?

File: Resources/Storage/GFAL2_StorageBase.py [hunk 4]
```diff
@@ -267,37 +257,37 @@ def __putSingleFile( self, src_file, dest_url, sourceSize ):
...
+        errStr = "GFAL2_StorageBase.__putFile: The local source file does not exist or is a directory"
+        self.log.error( errStr, src_file )
+        return S_ERROR( errno.ENOENT, errStr )
+      if not src_file.startswith( 'file://' ):
+        src_url = 'file://%s' % os.path.abspath( src_file )
+      sourceSize = getSize( src_file )
+      if sourceSize == -1:
+        errStr = "GFAL2_StorageBase.__putFile: Failed to get file size"
+        self.log.error( errStr, src_file )
+        return S_ERROR( errStr )
```
Comment (fstagni) [inline Resources/Storage/GFAL2_StorageBase.py]: error code?

File: Resources/Storage/GFAL2_StorageBase.py [hunk 5]
```diff
@@ -267,37 +257,37 @@ def __putSingleFile( self, src_file, dest_url, sourceSize ):
...
+      if sourceSize == 0:
+        errStr = "GFAL2_StorageBase.__putFile: Source file size is zero."
+        self.log.error( errStr, src_file )
+        return S_ERROR( errStr )
```
Comment (fstagni) [inline Resources/Storage/GFAL2_StorageBase.py]: error code?

File: Resources/Storage/GFAL2_StorageBase.py [hunk 6]
```diff
@@ -965,19 +957,17 @@ def __getChecksum( self, path, checksumType = None ):
...
+      return S_OK( fileChecksum )
 
-      except gfal2.GError, e:
-        errStr = 'GFAL2_StorageBase.__getChecksum: failed to calculate checksum.'
-        self.log.error( errStr, e.message )
-        return S_ERROR( e.message )
+    except gfal2.GError as e:
+      errStr = 'GFAL2_StorageBase.__getChecksum: failed to calculate checksum.'
+      self.log.error( errStr, repr( e ) )
+      return S_ERROR( e.code, repr( e ) )
```
Comment (fstagni) [inline Resources/Storage/GFAL2_StorageBase.py]: Previously you also used `return S_ERROR( e.code, e.message )` and here `return S_ERROR( e.code, repr( e ) )`.

What's better?

**Comment by fstagni** [comment]
There's no test? 

I see that in Resources/Storage/test/ there are several tests, but as of today they are not run "automatically" because the name of the test files is not "discoverable" by nosetests and py.test. They also seem more like integration tests rather than unit tests, but some expert (you? :) ) would need to look into it.

File: Resources/Storage/GFAL2_StorageBase.py (hunk 6 above)
Comment (chaen) [inline Resources/Storage/GFAL2_StorageBase.py]: Repr also contains the error number for display. I like it better

**Comment by chaen** [comment]
Corrected taking the comments into account. 
I wrote the Fle and XROOT plugin tests, I have no clue what the rest is :-) 
There are integration tests for gfal in what used to be TestDIRAC

**Comment by fstagni** [comment]
So, please rename those to "test_myTest.py" so that they are discovered. Also, I just tried to run the File plugin test in v6r15 and all failed (I could not verify the XRooT one)

Those tests that were in TestDIRAC are now in DIRAC/tests

**Comment by chaen** [comment]
The fixes are already in a PR waiting since October... I backport them here.

**Comment by fstagni** [comment]
Ok thanks. Now, please remove the `parseCommandLine` from these test files, as this should not be needed for unit tests.

**Comment by chaen** [comment]
As I said, I have no clue what these are supposed to do, and I am not even sure Andrei adapted them when he changed the base plugins, so they probably need a rewrite altogether, but I am not going to do it in this PR

**Comment by fstagni** [comment]
Review OK

File: Resources/Storage/StorageElement.py [hunk 7]
```diff
@@ -400,28 +400,38 @@ def getLocalPlugins( self ):
       return S_ERROR( self.errorReason )
     return S_OK( self.localPlugins )
 
-  def getStorageParameters( self, plugin ):
+  def getStorageParameters( self, plugin = None, protocol = None ):
```
Comment (atsareg) [inline Resources/Storage/StorageElement.py]: Is it necessary to keep both plugin and protocol ? Just Plugin is not enough ?

Comment (chaen) [inline Resources/Storage/StorageElement.py]: No, for srm for example. You don't know whether the plugin that will take care is SRM2 or GFAL2_SRM2

**Pull request merged**